### PR TITLE
Matlab performance improvements

### PR DESCRIPTION
**Pull request by bentoi**
This PR provides various performance improvements for decoding Slice classes:
- the generated code for classes handles no-init constructors to avoid a potentially costly initialisation of data members with default values
- replaces the use of `containers.Map` with array when possible
- many other small optimizations

### CODE REVIEW

File: cpp/src/slice2matlab/Main.cpp [hunk 1]
```diff
@@ -1744,7 +1752,19 @@ CodeVisitor::visitClassDefStart(const ClassDefPtr& p)
             //
             // Constructor
             //
-            if(!allMembers.empty())
+            if(allMembers.empty())
+            {
+                out << nl << "function " << self << " = " << name << spar << "noInit" << epar;
+                out.inc();
+                out << nl << "if nargin == 1 && ~strcmp(noInit, 'NoInit')";
```
Comment (pepone) [inline cpp/src/slice2matlab/Main.cpp:1759]: The problem is that if you have a class with a single string member, you cannot assign "NoInit" to the data member using the constructor,  I guess there is little chance for this to happen, but is kind of hacky.

**Review by pepone — APPROVED** [review]

File: matlab/lib/+IceInternal/ValueFactoryManagerI.m [hunk 2]
```diff
@@ -27,10 +28,14 @@ function add(obj, factory, id)
...
-            elseif obj.hasFactories && obj.factories.isKey(id)
-                r = obj.factories(id);
+            if obj.hasFactories
+                if isempty(id)
+                   r = obj.defaultFactory;
+                elseif obj.factories.isKey(id)
+                    r = obj.factories(id);
+                else
+                    r = obj.defaultFactory;
+                end
```
[the remark this answers is not in the document]
Comment (bentoi) [reply matlab/lib/+IceInternal/ValueFactoryManagerI.m:38]: Right.

File: matlab/lib/+IceInternal/EncapsDecoder.m [hunk 3]
```diff
@@ -67,91 +68,89 @@ function finish(obj)
...
             % Check if we have already unmarshalled the instance. If that's the case,
             % just invoke the callback and we're done.
             %
-            if obj.unmarshaledMap.isKey(index)
-                v = obj.unmarshaledMap(index);
-                cb(v);
-                return;
+            if index <= length(obj.unmarshaledMap)
+                v = obj.unmarshaledMap{index};
+                if ~isempty(v)
```
[the remark this answers is not in the document]
Comment (bentoi) [reply matlab/lib/+IceInternal/EncapsDecoder.m:134]: Yes.

File: cpp/src/slice2matlab/Main.cpp (hunk 1 above)
Comment (bentoi) [reply cpp/src/slice2matlab/Main.cpp]: Right. I don't remember why I used a string there but I've pushed a fix to use a `IceInternal.NoInit.Instance` constant instead.

**Pull request merged**